### PR TITLE
Add docker integration, CI, and internal `elm-review`

### PR DESCRIPTION
**Pull request by jiegillet**
This PR is meant to address everything left to do before deployment.

## Docker

It took me a while to figure out how docker works and how to debug it (especially since both `elm` and `elm-review` don't have explicit offline modes), but thanks to useful advice from Slack, I think I managed to get to something that works. I ended up with something very similar to `elm-test-runner`.
Of course, please do review carefully since this is my first docker build.

I made various changes to the shell scripts so that the same ones can run both locally and in docker.

## CI

I copied `docker.yml` pretty much as is (I only removed the `master` branch) from an Elixir repo, that should be fine. It doesn't run on my fork, probably @ErikSchierboom has to do some admin magic on it?

`elm_tests.yml` runs all the tests with `elm-test` (not `elm-test-rs` because version 1.1 doesn't allow tests with the same name, but IMO they make sense here. I didn't check later versions), checks the formatting, runs `elm-review` on the repo, and runs `run-tests-in-docker.sh`, which is basically `bin/smoke_test.sh` in the container.

## `elm-review`

I took the approach of (quote from README)

> As a rule of thumb, if we expect the students to follow specific rules when writing their solutions, we should hold ourselves to the same standards in our source code.

So I would like to keep the rules from the repo the same as the common rules we use for the students (we don't have any active at the moment, but I'm expecting we will add `NoUnused` and `Simplify` at some point).

All changes from the elm code was suggested from those rules.

## README

I wrote one.

## Going live

We will need to [merge that PR](https://github.com/exercism/website-copy/pull/2161) so we can check if the analyzer works as intended. You will get a comment if you use either `List.filter` or `List.filterMap`.

## Maintaining the repo

At this point, I would like to request to become a maintainer on the Elm track, if I can get the blessing of @mpizenberg and @ceddlyburge. 

I changed the `CODEOWNER` file so that we don't need admins to review each PR that don't touch the `.github` repo.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1 +1 @@
-* @exercism/maintainers-admin
+.github/CODEOWNERS @exercism/maintainers-admin
```
Comment (ErikSchierboom) [inline .github/CODEOWNERS:1]: Could you revert this change? We purposefully want an admin to sign off on any tooling.

File: .github/workflows/elm_tests.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f
+      
+      - name: Setup Elm
+        uses: jorelali/setup-elm@v3
```
Comment (ErikSchierboom) [inline .github/workflows/elm_tests.yml:14]: We have a policy of referring to workflows via their git SHA instead of a version number, as using a version number (tag) is a potential security vulnerability (the tag can be reassigned to a different commit). As this is an external dependency, this is especially important. Could you update?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,50 @@
+FROM node:lts-alpine AS builder
+
+# Working directory as specified by exercism
+WORKDIR /opt/analyzer
+
+# Copy source code
+COPY . .
```
Comment (ErikSchierboom) [inline Dockerfile:7]: I would move this until after the curl commands, as the latter are less likely to change than the source code and thus docker layer caching can be used better.

File: bin/run-in-docker.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+OUTPUT_DIR=$(realpath $3)
+
+# build docker image
+docker build -t elm-analyzer .
+
+# run image passing the arguments
+mkdir -p "$OUTPUT_DIR"
+docker run \
+    --rm \
+    --network none \
```
Comment (ErikSchierboom) [inline bin/run-in-docker.sh:45]: Just a FYI: we currently don't yet run on a readonly filesystem, but this might change in the future. You don't have to change anything, but just wanted to point it out.

File: test_data/strain/imperfect_solution/expected_analysis.json [hunk 5]
```diff
@@ -0,0 +1,2 @@
+{"summary":"Check the comments for things to fix. 🛠 ","comments":[{"comment":"elm.strain.do_not_use_filter","type":"essential","params":{}}]}
```
Comment (ErikSchierboom) [inline test_data/strain/imperfect_solution/expected_analysis.json:1]: It's important to have https://github.com/exercism/website-copy/pull/2161/files merged before the `analyzer` property is set to `true` in the track's `config.json` file: https://github.com/exercism/elm/blob/main/config.json#L19

File: test_data/strain/perfect_solution/expected_analysis.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{"summary":"No suggestions found.","comments":[]}
```
Comment (ErikSchierboom) [inline test_data/strain/perfect_solution/expected_analysis.json:1]: I'm wondering if the student benefits from having this summary being displayed, as just having

```suggestion
{"comments":[]}
```

would have the website show that there are no suggestions. I think I would prefer the latter.

**Review by ErikSchierboom — COMMENTED** [review]
Great stuff! I've left some comments.

@iHiD Could you give this repo access to the proper secrets?

**Comment by ceddlyburge** [comment]
> At this point, I would like to request to become a maintainer on the Elm track, if I can get the blessing of @mpizenberg and @ceddlyburge.

You have my blessing :)

**Comment by mpizenberg** [comment]
Mine too :)

File: Dockerfile (hunk 3 above)
Comment (jiegillet) [reply Dockerfile]: Oh, that's a clever advice, thank you!

File: bin/run-in-docker.sh (hunk 4 above)
Comment (jiegillet) [reply bin/run-in-docker.sh:45]: What does readonly mean exactly? Not being able to create any sort of file anywhere on the container?
Because if that's the case, it would become a lot more tricky to make the analyzer or even test runner work...

File: test_data/strain/imperfect_solution/expected_analysis.json (hunk 5 above)
Comment (jiegillet) [reply test_data/strain/imperfect_solution/expected_analysis.json:1]: Noted! I didn't know about that line in`config.json`, thank you.

I don't think I can request a review from @mpizenberg or @ceddlyburge on website-copy, they are not on the list. How can we add them in there? We will need to approve analyzer comments in the future. Maybe in the mean time you could approve it?

File: test_data/strain/perfect_solution/expected_analysis.json (hunk 6 above)
Comment (jiegillet) [reply test_data/strain/perfect_solution/expected_analysis.json:1]: I'm pretty sure that when there are no comments, the website doesn't display anything.
The elixir analyzer exports the exact same file and nothing is shown.
I'd prefer having the summary there, even if it isn't displayed, it makes the code simpler and the analysis file simpler to understand from a dev point of view.

Screen Shot 2022-03-02 at 10 42 03

**Review by jiegillet — COMMENTED** [review]
Thank you very much.
I think I've addressed all the comments.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,53 @@
+FROM node:lts-alpine AS builder
```
Comment (ceddlyburge) [inline Dockerfile:1]: I use a tool called [HadoLint](https://github.com/hadolint/hadolint) in VsCode to lint Dokerfiles. It has 3 warning messages and 1 info message for this file. There is nothing drastic but it is probably worth taking a look. If you can't run the tool yourself let me know and I can drop the messages in here or we can have a call.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+## Tests
+
+An overall check of the analyzer can be ran with
+```shell
+./bin/smoke_test.sh
+```
+
+This will run the Analyzer on a predefined set of exercises in `test_data` and compare the outputs to expected outputs.
+
+More tests are defined in the `tests` folder and can be ran with `elm-test`.
```
Comment (ceddlyburge) [inline README.md:46]:
```suggestion
More tests are defined in the `tests` folder and can be run with `elm-test`.
```

File: bin/run-in-docker.sh (hunk 4 above)
Comment (mpizenberg) [reply bin/run-in-docker.sh:45]: No network is fine though, my remark was just regarding readonly

File: README.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
+# Elm Analyzer
```
Comment (ceddlyburge) [inline README.md:1]: Excellent readme ...

File: bin/check_files.sh [hunk 10]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/sh
```
Comment (ceddlyburge) [inline bin/check_files.sh:1]: Is this part of the smoke test? It might be worth making this obvious in some way if so.

**Review by ErikSchierboom — APPROVED** [review]

File: test_data/strain/imperfect_solution/expected_analysis.json (hunk 5 above)
Comment (mpizenberg) [reply test_data/strain/imperfect_solution/expected_analysis.json:1]: Indeed, I'd prefer not to be added to website-copy. I don't mind occasionally being pingged there for analyzer messages though, even if I cannot officially review. I might just comment if I find the time, or not if I don't.

Comment (ErikSchierboom) [reply test_data/strain/imperfect_solution/expected_analysis.json:1]: I've just merged the website-copy PR

File: README.md (hunk 9 above)
Comment (jiegillet) [reply README.md:1]: Thank you :)

File: Dockerfile (hunk 7 above)
Comment (jiegillet) [reply Dockerfile:1]: I managed to install it and make it happy :)
Thank you for the recommendation.

**Comment by jiegillet** [comment]
I know I was told not to worry, but I modified the code so that it would run in read-only (but still writing in `/tmp`).

Is there anything left to do before generating the secrets and activating the analyzer in the Elm config file?

**Comment by mpizenberg** [comment]
> I know I was told not to worry, but I modified run.sh and the smoke test so that it would run in read-only

Haha It's like you just climbed a mountain, you have a great view, but there is this last spot, 50m away, and your journey would not be complete until you make sure the view isn't better from up there.

If I have a small nitpick (without checking in detail all the PR), why are some files are with an eol at the eof while some do not have an eol? I think the posix standard ask that we add an eol as the last char of a file. If you use VSCode it should be this setting: "files.insertFinalNewline": true

**Comment by jiegillet** [comment]
Haha yeah, it's that kind of feeling. And when you reach this other perfect peak, you spot another one :)

Thank you for the EOL tip, it is something that I never really paid attention to, I do use VSCode and I've enabled the setting.

File: .github/workflows/docker.yml [hunk 11]
```diff
@@ -0,0 +1,21 @@
+name: Deploy
```
Comment (ErikSchierboom) [inline .github/workflows/docker.yml:1]: This file is now a duplicate of the `deploy.yml` file and can be removed.

**Comment by ErikSchierboom** [comment]
Form my point of view, this is ready to be merged!

**Comment by ceddlyburge** [comment]
me too :)
